bug in interface usage in combination with or block

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Pop!_OS 21.10
Processor: 6 cpus, 64bit, little endian, AMD Ryzen 5 4500U with Radeon Graphics
CC version: cc (Ubuntu 11.2.0-7ubuntu2) 11.2.0

getwd: /home/astro/Projects/V/scratch/isolate
vmodules: /home/astro/.vmodules
vroot: /home/astro/Projects/V/v-main
vexe: /home/astro/Projects/V/v-main/v
vexe mtime: 2022-05-06 07:22:06
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 8a18f91.45fe87c

Git version: git version 2.32.0
Git vroot status: weekly.2022.16-159-g45fe87c9
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 eee03626

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg main.v`
```v
module main

import rand

interface Sample {
mut:
	get_next() int
}

struct SampleA {
mut:
	state int
}

fn (mut sample SampleA) get_next() int {
	sample.state++
	return sample.state
}

struct SampleB {
mut:
	state int = 1
}

fn (mut sample SampleB) get_next() int {
	sample.state += 2
	return sample.state
}

fn create_sampler() Sample {
	return if rand.intn(2) or { 0 } == 0 { Sample(SampleA{}) } else { Sample(SampleB{}) }
}

fn main() {
	mut sample := create_sampler()
	println(sample.get_next())
	println(sample.get_next())
}

```

**What did you expect to see?**

Normal compilation and working program outputing (1,2) or (3,5)


**What did you see instead?**
```
==================
/tmp/v_1000/main.16381773999166518543.tmp.c:16969: error: ';' expected (got "}")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -5133,9 +5133,7 @@ fn (mut g Gen) or_block(var_name string, or_block ast.OrExpr, return_type ast.Ty
 					g.inside_opt_data = true
 					g.expr_with_cast(expr_stmt.expr, expr_stmt.typ, return_type.clear_flag(.optional))
 					g.inside_opt_data = old_inside_opt_data
-					if g.inside_ternary == 0 {
-						g.writeln(';')
-					}
+					g.writeln(';')
 					g.stmt_path_pos.delete_last()
 				} else {
 					g.stmt(stmt)

--- a/vlib/v/tests/if_cond_with_optional_test.v
+++ b/vlib/v/tests/if_cond_with_optional_test.v
@@ -1,0 +1,43 @@
+module main
+
+import rand
+
+interface Sample {
+mut:
+	get_next() int
+}
+
+struct SampleA {
+mut:
+	state int
+}
+
+fn (mut sample SampleA) get_next() int {
+	sample.state++
+	return sample.state
+}
+
+struct SampleB {
+mut:
+	state int = 1
+}
+
+fn (mut sample SampleB) get_next() int {
+	sample.state += 2
+	return sample.state
+}
+
+fn create_sampler() Sample {
+	return if rand.intn(1) or { 0 } == 0 { Sample(SampleA{}) } else { Sample(SampleB{}) }
+}
+
+fn test_if_cond_with_optional() {
+	mut sample := create_sampler()
+	mut ret := sample.get_next()
+	println(ret)
+	assert ret == 1
+
+	ret = sample.get_next()
+	println(ret)
+	assert ret == 2
+}

